posit-dev/publisher
# File Credentials Support

## Intent

Full support for credentials persistence via a file `~/.connect-credentials`.

Following up the work done at #2308, this PR completes the feature by:
- Having a main credentials service constructor `NewCredentialsService`
- `NewCredentialsService` is called by endpoints that do credentials operations
- `NewCredentialsService` tries to:
    -  set up the default system's keyring service, `NewKeringCredentialsService`
    - falling back to file based credentials when there is no keyring available. `NewFileCredentialsService`
- This allows for different funtionality on credentials persistence to share the same interface for credentials operations `List(), Get(), Set(), Delete()`

**Screenshots of Workbench using the file backed credentials service**
<img width="2190" alt="Screen Shot 2024-09-25 at 15 05 32" src="https://github.com/user-attachments/assets/38e72d17-e798-475f-8042-87caffc422a5">

<img width="910" alt="Screen Shot 2024-09-25 at 15 11 30" src="https://github.com/user-attachments/assets/d8526830-645f-4122-a663-43c0bd04382d">

## Type of Change

- - [ ] Bug Fix 
- - [x] New Feature 
- - [ ] Breaking Change 
- - [ ] Documentation 
- - [ ] Refactor 
- - [ ] Tooling 

## Automated Tests

Unit tests created and updated

## Directions for Reviewers

To manually verify this changes: 
- Start a new workbench session,
- Install the compatible packaged VSIX file, generated by CI on this PR's job
- Reload session if needed, from the command pallette `> Developer: Reload Window`
- Create and delete credentials as desired
- Remove workbench sessions, see how credentials are persisted next time you start a new session

## Checklist
